int_scan.cpp

```diff
@@ -0,0 +1,115 @@
+#include "int_scan.hpp"
+#include <cassert>
+#include <ctype.h>
+
+static const int OCT = 8;
+static const int DEC = 10;
+static const int HEX = 16;
+
+/*
+The function returns a member of the Base enumeration, which represents the base
+of the input string. If the input string starts with the character 0 (zero),
+then the input string represents an octal number. If the input string starts
+with the characters 0x or 0X (zero x/X), then the input string represents a
+hexadecimal number. If the input string does not have one of these prefixes,
+then the input string represents a decimal number. The function starts looking
+for the prefix from the index specified by the position pointer. The function
+also returns the position of the next character of the input to be read after
+the prefix (if it exists).
+*/
+static int check_base(const char *input, int *position) {
+  int pos = 0;
+  if (input[pos] != '0') {
+    *position = pos;
+    return DEC;
+  }
+  pos++;
+  if (tolower(input[pos]) == 'x') {
+    pos++;
+    *position = pos;
+    return HEX;
+  }
+  *position = pos;
+  return OCT;
+}
```

Comment:
I think it's better if you transform this to:
```cpp
  int ret;
  int pos = 0;
  if (input[pos] != '0') {
    ret = DEC;
  } else if (tolower(input[pos+1]) == 'x') {
    pos += 2;
    ret = HEX;
  } else {
    ++pos;
    ret = OCT;
  }
  *position = pos;
  return ret;
```
You can even be "smarter" by observing that for every base, there's a different advance in position.
To put it otherwise, every base is identified by its advance. So, this function could be renamed to `advance_pos()`
and the rest of the code work like:
```cpp
int advance = advance_pos(input);
if (advance == 0) // DEC
else if (advance == 1) // OCT
...
input = input + advance;
```
It's a good way to avoid passing a pointer but I'm not sure whether I recommend that :p